File: project RAD/BFS/autocomplete system proble,/sol.py
from typing import List, Dict, Optional, Tuple
import collections

# --- 1. Trie Node Structure ---
class TrieNode:
    def __init__(self):
        # Maps character to the next TrieNode
        self.children: Dict[str, TrieNode] = collections.defaultdict(TrieNode)
        # Stores hit count for sentences ending at this node.
        self.counts: collections.defaultdict[str, int] = collections.defaultdict(int)

# --- 2. Autocomplete System Class ---
class AutocompleteSystem:
    def __init__(self, sentences: List[str], times: List[int]):
        self.root = TrieNode()
        self.current_query: str = ""
        self.current_node: Optional[TrieNode] = self.root
        
        # Build the initial Trie
        for s, t in zip(sentences, times):
            self._insert(s, t)

    def _insert(self, sentence: str, time: int):
        """Helper to insert a sentence into the Trie."""
        node = self.root
        for char in sentence:
            node = node.children[char]
        # Store the sentence and update its count at the final node
        node.counts[sentence] += time

    def _dfs_search(self, node: TrieNode) -> List[Tuple[int, str]]:
        """Helper to collect all (count, sentence) pairs reachable from a node."""
        results = []
        # DFS to traverse all children
        for child_node in node.children.values():
            results.extend(self._dfs_search(child_node))
        
        # Collect results from the current node's ending sentences
        for sentence, count in node.counts.items():
            results.append((-count, sentence)) # Use -count for Min-Heap/Min-Sort behavior
        
        return results

    def input(self, c: str) -> List[str]:
        # Case 1: End of Query
        if c == '#':
            if self.current_query:
                # Update the count for the completed sentence
                self._insert(self.current_query, 1)
            
            # Reset state for the next query
            self.current_query = ""
            self.current_node = self.root
            return []
        
        # Case 2: Continuing the Query
        self.current_query += c
        
        # If the last character led to a dead end, all future searches will be empty
        if self.current_node is None:
            return []

        # Traverse to the next node
        if c in self.current_node.children:
            self.current_node = self.current_node.children[c]
        else:
            # Dead end reached
            self.current_node = None
            return []

        # Find all sentences starting with the current query using DFS
        # We need the results from the current node AND all its children
        results = []
        if self.current_node:
            results.extend(self._dfs_search(self.current_node))

        # Add sentences that end *at* the current node's position
        for sentence, count in self.current_node.counts.items():
            results.append((-count, sentence))
        
        # Sort results: first by count (desc), then by sentence (alpha asc)
        # Sorting by tuple automatically handles this: (-count, sentence)
        results.sort()
        
        # Return the top 3 sentences
        return [s for count, s in results[:3]]

# --- Test Cases ---

if __name__ == "__main__":
    
    sentences = ["i love you", "i love python", "i like coffee", "best coffee ever"]
    times = [5, 3, 2, 8]
    
    acs = AutocompleteSystem(sentences, times)
    
    print("--- Autocomplete System Test ---")
    
    # 1. Start typing 'i'
    result1 = acs.input('i')
    # Expected: ["i love you", "i love python", "i like coffee"]
    print(f"Input 'i': {result1}")
    
    # 2. Continue with ' '
    result2 = acs.input(' ')
    # Expected: ["i love you", "i love python", "i like coffee"]
    print(f"Input ' ': {result2}")
    
    # 3. Continue with 'l' (i l)
    result3 = acs.input('l')
    # Expected: ["i love you", "i love python", "i like coffee"]
    print(f"Input 'l': {result3}")

    # 4. Finish and update count
    result4 = acs.input('#')
    # Expected: [] (The sentence "i l" is now recorded with count 1)
    print(f"Input '#': {result4}")

    # 5. Type a new popular prefix 'b'
    result5 = acs.input('b')
    # Expected: ["best coffee ever"]
    print(f"Input 'b': {result5}")

File: project RAD/BFS/autocomplete system proble,/sol.java
import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    Map<String, Integer> counts = new HashMap<>();
}

class AutocompleteSystem {
    private TrieNode root;
    private TrieNode currentNode;
    private StringBuilder currentQuery;

    public AutocompleteSystem(String[] sentences, int[] times) {
        root = new TrieNode();
        currentNode = root;
        currentQuery = new StringBuilder();

        for (int i = 0; i < sentences.length; i++) {
            insert(sentences[i], times[i]);
        }
    }

    private void insert(String sentence, int time) {
        TrieNode node = root;
        for (char c : sentence.toCharArray()) {
            node.children.putIfAbsent(c, new TrieNode());
            node = node.children.get(c);
        }
        node.counts.put(sentence, node.counts.getOrDefault(sentence, 0) + time);
    }

    private List<Map.Entry<String, Integer>> dfsSearch(TrieNode node) {
        List<Map.Entry<String, Integer>> result = new ArrayList<>();
        result.addAll(node.counts.entrySet());

        for (TrieNode child : node.children.values()) {
            result.addAll(dfsSearch(child));
        }

        return result;
    }

    public List<String> input(char c) {
        // Case 1: End of sentence
        if (c == '#') {
            insert(currentQuery.toString(), 1);
            currentQuery = new StringBuilder();
            currentNode = root;
            return new ArrayList<>();
        }

        // Case 2: Add character to query
        currentQuery.append(c);

        if (currentNode == null)
            return new ArrayList<>();

        if (!currentNode.children.containsKey(c)) {
            currentNode.children.put(c, new TrieNode());
            currentNode = currentNode.children.get(c);
            return new ArrayList<>();
        }

        currentNode = currentNode.children.get(c);
        List<Map.Entry<String, Integer>> allResults = dfsSearch(currentNode);

        // Sort by count desc, then alphabetically
        allResults.sort((a, b) -> {
            if (!a.getValue().equals(b.getValue())) {
                return b.getValue() - a.getValue();
            }
            return a.getKey().compareTo(b.getKey());
        });

        List<String> top3 = new ArrayList<>();
        for (int i = 0; i < Math.min(3, allResults.size()); i++) {
            top3.add(allResults.get(i).getKey());
        }

        return top3;
    }

    public static void main(String[] args) {
        String[] sentences = { "i love you", "i love python", "i like coffee", "best coffee ever" };
        int[] times = { 5, 3, 2, 8 };

        AutocompleteSystem acs = new AutocompleteSystem(sentences, times);

        System.out.println("Input: i -> " + acs.input('i'));
        System.out.println("Input:  (space) -> " + acs.input(' '));
        System.out.println("Input: l -> " + acs.input('l'));
        System.out.println("Input: # -> " + acs.input('#'));
        System.out.println("Input: b -> " + acs.input('b'));
    }
}
